### PR TITLE
Connections better use typedefed types Segment, Synapse, ...

### PR DESCRIPTION
**Pull request by breznak**
- Connections (and related SP, TM) properly use typedefed types for Segment, SynapseIdx etc
  - allowing us to really change them as needed
- minor cleanups

Implements size_t for TM, SP, Connections #357

### CODE REVIEW

File: src/nupic/algorithms/Connections.cpp [hunk 1]
```diff
@@ -303,10 +306,9 @@ void Connections::updateSynapsePermanence(Synapse synapse,
       potentialPreseg.push_back( segment );
     }
 
-    for (auto h : eventHandlers_) {
+    for (auto h : eventHandlers_) { //TODO handle callbacks in performance-critical method only in Debug?
```
Comment (breznak) [inline src/nupic/algorithms/Connections.cpp]: this method is a hotspot I found in profiling, do we really need the callbacks here, can be left out only for debug,.. ?

File: src/nupic/algorithms/Connections.hpp [hunk 2]
```diff
@@ -41,11 +41,11 @@ namespace nupic {
 namespace algorithms {
 namespace connections {
 
-typedef UInt32 CellIdx;
-typedef UInt16 SegmentIdx; /** Index of segment in cell. */
-typedef UInt16 SynapseIdx; /** Index of synapse in segment. */
-typedef UInt32 Segment;    /** Index of segment's data. */
-typedef UInt32 Synapse;    /** Index of synapse's data. */
+typedef UInt32 CellIdx;  //TODO instead of typedefs, use templates for proper type-checking
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: typedefs have the disadvantage of only renaming the type for the user(=programmer), for proper type-checking (so that compiler checks difference for SegmentIdx, SynapseIdx) we'd need to use templates instead. 
I'll consider that in subsequent PR.

File: src/nupic/algorithms/Connections.hpp [hunk 3]
```diff
@@ -41,11 +41,11 @@ namespace nupic {
...
+typedef unsigned char SegmentIdx; /** Index of segment in cell. */
+typedef unsigned char SynapseIdx; /** Index of synapse in segment. */
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: @ctrl-z-9000-times I replaced lot of plain UInts with these typedefs, can you please double-check I use them as intended? (ei SynapseIdx and not Synapse, etc..)

File: src/nupic/algorithms/Connections.hpp [hunk 4]
```diff
@@ -41,11 +41,11 @@ namespace nupic {
...
+typedef UInt16 Segment;    /** Index of segment's data. */
+typedef UInt16 Synapse;    /** Index of synapse's data. */
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: note, a WIP commit changes these to much lower values. 
I'll need to asses the impact, suprisingly, the lower values (compared to UInt) do not gain a significant performance gain (possibly some forgotten type-casting happens?) 

Unit test SP:: ExactOutput "breaks" as stays stuck, does not finish, nor fail. I don;t know why

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: The size of this data type controls the maximum number of synapses & segments.
So with an 8-bit integer, it can only have a 255 synapses, maximum.
16-bits as shown here allows a maximum of 65,536 synapses.

I will give this a more thorough review within a day or two.

File: src/nupic/algorithms/Connections.cpp (hunk 1 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.cpp]: Yea I have no idea why these exist but their part of the public API so I didn't remove them.
Maybe try: `for( const auto & : eventHandlers_)` for speed?

File: src/nupic/algorithms/Connections.cpp [hunk 5]
```diff
@@ -593,7 +596,7 @@ void Connections::load(std::istream &inStream) {
         inStream >> destroyedSegment;
       }
 
-      Segment segment = {(UInt32)-1};
+      Segment segment = {std::numeric_limits<Segment>::max()-1};
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.cpp]: -1 == std::numeric_limits::max()
You've got an extra -1.

File: src/test/unit/algorithms/SpatialPoolerTest.cpp [hunk 6]
```diff
@@ -70,6 +70,13 @@ bool check_vector_eq(UInt arr[], vector<UInt> vec) {  //TODO replace with ArrayB
   return true;
 }
 
+bool check_vector_eq(UInt arr[], vector<SynapseIdx> vec) {
```
Comment (ctrl-z-9000-times) [inline src/test/unit/algorithms/SpatialPoolerTest.cpp]: Where is this used?

File: src/nupic/algorithms/Connections.hpp (hunk 3 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: I read through this and I think you've got all of the types right.

You need to add checks to the methods `Connections::addSynapse` & `Connections::addSegment` to check for too many synapses/segments.  Otherwise it will fail silently.

File: src/nupic/algorithms/Connections.cpp (hunk 1 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: I'll profile w/o the loop, imo this would be a candidate to break the API, if shows performance gain.

File: src/nupic/algorithms/Connections.hpp (hunk 3 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: > You need to add checks to the methods Connections::addSynapse & Connections::addSegment to check for too many synapses/segments. Otherwise it will fail silently. 

Great point! I had only the option to detect overflow in Debug

File: src/nupic/algorithms/Connections.hpp (hunk 4 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: > So with an 8-bit integer, it can only have a 256 synapses, maximum.
16-bits as shown here allows a maximum of 65,536 synapses.

yes. Are you suggesting the 16b should match the 8-bit? As SynapseIdx is "synapses per segment", and `Synapse` is a flat array/index to all existing synapses (their data). But there can be a number of segments in cells, so Synapses >> SynapseIdx.

File: src/nupic/algorithms/Connections.cpp (hunk 5 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: might have been the cause why the test stalled

File: src/test/unit/algorithms/SpatialPoolerTest.cpp (hunk 6 above)
Comment (breznak) [reply src/test/unit/algorithms/SpatialPoolerTest.cpp]: the SP Test uses a number of variations of those. I added a TODO to improve that, but for now I only add another declaration for the version used in these tests.

File: src/nupic/algorithms/Connections.hpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: > Are you suggesting

No i just wanted to clarify.
What u have here looks good, as long as it doesnt overflow.

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: And if it does overflow then we can easily fix it by changing these typedefs which after this PR should be used much more consistently. :)

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 7]
```diff
@@ -450,17 +461,16 @@ static void punishPredictedColumn(
...
 
 void TemporalMemory::activateCells(const size_t activeColumnsSize,
                                    const UInt activeColumns[], bool learn) {
   if (checkInputs_ && activeColumnsSize > 0) {
-    NTA_CHECK(isSortedWithoutDuplicates(activeColumns,
-                                        activeColumns + activeColumnsSize-1))
-        << "The activeColumns must be a sorted list of indices without "
-           "duplicates.";
+    NTA_CHECK(std::is_sorted(activeColumns, activeColumns + activeColumnsSize-1))
+        << "The activeColumns must be a sorted list of indices without duplicates.";
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: @ctrl-z-9000-times with regards to #337 , I'm having problems with this code (check), 
calling this from the SDR, the list is never sorted, so I have to cast to vector, sort, and pass to the old-API version. 
Solutions: 
- SDR.getSparse() would return a sorted vector of indices (slower first, but cached) 
- remove this requirement (and check) from activateCells (preferable) 
thoughts?

Comment (dkeeney) [reply src/nupic/algorithms/TemporalMemory.cpp]: I don't know how much about how sparse data is used in the algorithms but if it is sorted then algorithms can can do merge style processing (single pass) when doing operations on two sparse vectors.
See VectorHelpers::unionOfVectors( )

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: You should be able to sort the data in place. GetSparse returns a non-constant reference.
`std::sort( sdr.getSparse().begin(), sdr.getSparse().end() )`

Comment (dkeeney) [reply src/nupic/algorithms/TemporalMemory.cpp]: That would work.  if you remove the sorted requirement, you can add that sort to unionOfVectors( ).  I wonder if there is other code that assumes that sparse data is sorted?

Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: > remove this [sorted, unique] requirement (and check) from activateCells (preferable)

would not work, `iterGroupBy` later in that function requires this from the data! #263

**Comment by breznak** [comment]
This is now ready for your review. 
Let me know if Connections should use typedefs for specifying the data types (as is done now), 
or rather templates. 
https://github.com/htm-community/nupic.cpp/pull/331#discussion_r266506737 

Templates: 
- pros: 
  - better type checks
  - Connections can be instantiated with different backend types
- cons: 
  - will need to rewrite code, merge cpp+hpp (templates in c++ dont allow separate cpp)

**Comment by dkeeney** [comment]
The templates are a great tool BUT when used in the extreme it can have some nasty side-effects.  Cereal is getting close to that in my opinion. 

It is much like Macros.  The are an important feature to have around but if you nest them many layers deep the code becomes unmaintainable.  I have some war stories I could tell you about very bad experiences with macros like that.  That is why I am very hesitant about layers of templates as well.  

In this case we probably don't have the layers that Cereal does so this might not be an issue. But if it is a toss-up between typedefs and templates I recommend the typedefs.  The typedefs can also be abused but I don't think it is the case here.

File: src/nupic/algorithms/Connections.cpp [hunk 8]
```diff
@@ -358,9 +365,9 @@ const SynapseData &Connections::dataForSynapse(Synapse synapse) const {
   return synapses_[synapse];
 }
 
-UInt32 Connections::segmentFlatListLength() const { return (UInt32)segments_.size(); }
+Segment Connections::segmentFlatListLength() const { return (Segment)segments_.size(); }
```
Comment (dkeeney) [inline src/nupic/algorithms/Connections.cpp]: If I remember right, `Segment` is both a typedef and a class elsewhere.  I know the namespaces keep the compilers from getting confused but but that does not help people trying to remember what is what.  Or helping sort out the results from text searches.  I am not recommending any changes, just pointing this out.

**Comment by breznak** [comment]
> The templates are a great tool BUT when used in the extreme it can have some nasty side-effects. 

that's true. 
Well, the question "template x typedef" boils down to: will it make sense to have any 2 Connections instances at runtime with different backend types? 
And I'll leave that decision after benchmarking, which is now easily allowed.

File: src/test/unit/algorithms/SpatialPoolerTest.cpp [hunk 9]
```diff
@@ -2011,7 +2018,7 @@ TEST(SpatialPoolerTest, testConstructorVsInitialize) {
   EXPECT_TRUE(sp1 == sp2) << "Spatial Poolers not equal";
 }
 
-TEST(SpatialPoolerTest, ExactOutput) {
+TEST(SpatialPoolerTest, DISABLED_ExactOutput) { //FIXME
```
Comment (dkeeney) [inline src/test/unit/algorithms/SpatialPoolerTest.cpp]: Is this test case broken?

**Comment by ctrl-z-9000-times** [comment]
Im against templates as well.  I think theyre complicated and error prone. Before using them we should profile it to make sure they will actually improve it.

File: src/nupic/algorithms/Connections.cpp (hunk 8 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: yes, users should not be too confused either: 
- connections already operates (typedefs) a Segment, so it's that one (Segment here = UInt) 
- the other class Segment is only used in old TP,Cells4, and proposed for removal

File: src/test/unit/algorithms/SpatialPoolerTest.cpp (hunk 9 above)
Comment (breznak) [reply src/test/unit/algorithms/SpatialPoolerTest.cpp]: good spot, I'll have a look!

Comment (breznak) [reply src/test/unit/algorithms/SpatialPoolerTest.cpp]: enabled, thanks!

**Comment by ctrl-z-9000-times** [comment]
I see the following issue in several places:
If you have set stored as a `vector<UInt8>` then it can hold 256 elements, indexed as 0-255.  You can not represent the size of this vector as a UInt8, since 256 will overflow into 0.

One way to fix this is to lower the limits by one, so in this example the vector would have at most 255 elements which is representable as a UInt8.

Another way to fix this is to increase the size of the return type.  Internally the vector uses an `unsigned int` to keep sizes.

**Comment by breznak** [comment]
> If you have set stored as a vector<UInt8> then it can hold 256 elements, indexed as 0-255. You can not represent the size of this vector as a UInt8, since 256 will overflow into 0. 

@ctrl-z-9000-times  can you point me to an example, please? 
I've searched for `vector<UInt8>` and found none, also the smallest typedef used is unsigned 16b:  
https://github.com/htm-community/nupic.cpp/pull/331/files#diff-dff563d2b4cacf6184c85c53ed96cf75R46

File: src/nupic/algorithms/Connections.cpp [hunk 10]
```diff
@@ -83,11 +83,13 @@ void Connections::unsubscribe(UInt32 token) {
...
-  if (destroyedSegments_.size() > 0) {
+  if (!destroyedSegments_.empty() ) {
     segment = destroyedSegments_.back();
     destroyedSegments_.pop_back();
   } else {
-    segment = (Segment)segments_.size();
+    NTA_CHECK(segments_.size() <= std::numeric_limits<Segment>::max()) << "Add segment failed: Range of Segment (data-type) insufficinet size."
+	    << segments_.size() << " <= " << std::numeric_limits<Segment>::max();
+    segment = static_cast<Segment>(segments_.size());
     segments_.push_back(SegmentData());
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.cpp]: If before this call: segments_.size() == numeric_limits<Segment>::max()
Then after this call: segments_.size() > numeric_limits<Segment>::max()
Instead it should raise an exception.

File: src/nupic/algorithms/Connections.cpp [hunk 11]
```diff
@@ -628,20 +618,20 @@ void Connections::load(std::istream &inStream) {
 
 CellIdx Connections::numCells() const { return (CellIdx)cells_.size(); }
 
-UInt Connections::numSegments() const {
-  return (UInt)(segments_.size() - destroyedSegments_.size());
+Segment Connections::numSegments() const {
+  return (Segment)(segments_.size() - destroyedSegments_.size());
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.cpp]: Because of this typecase: `segments_.size()` must be `<= numeric_limits<Segment>::max()`

File: src/nupic/algorithms/Connections.cpp (hunk 10 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: thanks, I see. I should not allow push_back new segment if size() already == limits_max.

File: src/nupic/algorithms/Connections.cpp (hunk 11 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: here, using Segment is a bit confusing, should we rather use size_t for anything that returns a size of something? (and check it actually fits where used?)

Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: I'll improve this in #357

File: src/nupic/algorithms/Connections.cpp (hunk 10 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.cpp]: @ctrl-z-9000-times fixed the `<` and actually added tests that verify the overflow would be detected.

File: bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp [hunk 12]
```diff
@@ -53,14 +53,21 @@ using nupic::algorithms::connections::Permanence;
...
 
         py_HTM.def(py::init<>());
-        py_HTM.def(py::init<std::vector<UInt>
-                , UInt, UInt
-                , Permanence, Permanence
-                , UInt, UInt
-                , Permanence, Permanence, Permanence
+        py_HTM.def(py::init<std::vector<CellIdx>
+	        , CellIdx
+		, SynapseIdx
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: bindings too now use proper typedefs

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 13]
```diff
@@ -663,7 +672,7 @@ UInt TemporalMemory::columnForCell(const CellIdx cell) const {
   return cell / cellsPerColumn_;
 }
 
-vector<CellIdx> TemporalMemory::cellsForColumn(Int column) {
+vector<CellIdx> TemporalMemory::cellsForColumn(Int column) { //TODO remove, incorrect w/o topology
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: Off-topic:
@ctrl-z-9000-times TM does not use topology, so how can this be correct?
I would like if topology (and cellsToColumns()) could be moved to Connections, I think it's a generic feature that could be useful on other scenarios. 
Related ColumnPooler #293  (that defines Topology), Anomaly #347 (that needs to convert from cells to columns)

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 14]
```diff
@@ -733,7 +742,7 @@ vector<Segment> TemporalMemory::getMatchingSegments() const
   return matchingSegments_;
 }
 
-UInt TemporalMemory::numberOfColumns() const { return numColumns_; }
+CellIdx TemporalMemory::numberOfColumns() const { return numColumns_; }
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: this will be changed in #357

File: src/test/unit/algorithms/ConnectionsTest.cpp [hunk 15]
```diff
@@ -677,4 +677,39 @@ TEST(ConnectionsTest, testSaveLoad) {
   ASSERT_EQ(c1, c2);
 }
 
+TEST(ConnectionsTest, testCreateSegmentOverflow) {
+    const size_t LIMIT = std::numeric_limits<Segment>::max();
+    if(LIMIT <= 256) { //connections::Segment is too large (likely uint32), so this test would run, but memory
```
Comment (breznak) [inline src/test/unit/algorithms/ConnectionsTest.cpp]: these are the new tests that check overflow in createSynapse/Segment methods; 
but I cannot override at runtime the connections::Segment type (this is where teplates, instead of typedefs, would be useful), so the test is effectively skipped, unless the user manually changes the Segment/Synapse to unsigned char.

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 13 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: The word "columns" is used in two different contexts with two different meanings.

The TM & Anomaly have *mini*-columns.  Mini-Columns are not a generic feature, they should stay where they are (in the TM).

The SP & CP have *macro*-columns, which have to do with large scale topology.

Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: > The TM & Anomaly have mini-columns. Mini-Columns are not a generic feature, they should stay where they are (in the TM). 

and a mini-column: 
- uses same feed-forward input, has different horizontal connections to other cells
- uses naive topology: if TM has 8 cells with indices 1..8, and 4 cells-per-column: 
  then cells 1-4 are in mini-col 1, c 5-8 in mini-col 2; 
  - Q1 does it make sense to support more complex topologies? 
  - Q2 do we support nD TM? would it flatten nicely that neighbouring cells (in 3D) would have nearby indices (CellIdx)? 
- VectorHelpers::cellsToColumns() should be then moved to TM, as it makes sense only for TM's cells, not any vector/SDR.

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: Yes, that is correct.

>  Q1 does it make sense to support more complex topologies?

Not really. The TM does not have topology.  It has groups of cells which are organized into mini-columns. The mini-columns don't have any topology between them. The SP -not TM- is responsible for large scale topology.

> Q2 do we support nD TM? would it flatten nicely that neighbouring cells (in 3D) would have nearby indices (CellIdx)?

No.  The TM can work with nD inputs, by flattening them into 1D inputs.  This works as long as the SP has the correct topology and imparts it on the SDR which it gives to the TM.

>  VectorHelpers::cellsToColumns() should be then moved to TM, as it makes sense only for TM's cells, not any vector/SDR.

Yes.  Or better yet mixed with the anomaly code which will be baked into the TM.

File: src/nupic/algorithms/Connections.hpp [hunk 16]
```diff
@@ -41,12 +41,13 @@ namespace nupic {
...
 namespace connections {
 
-typedef UInt32 CellIdx;
-typedef UInt16 SegmentIdx; /** Index of segment in cell. */
-typedef UInt16 SynapseIdx; /** Index of synapse in segment. */
-typedef UInt32 Segment;    /** Index of segment's data. */
-typedef UInt32 Synapse;    /** Index of synapse's data. */
-typedef Real32 Permanence;
+//TODO instead of typedefs, use templates for proper type-checking?
+using CellIdx   = nupic::sdr::ElemSparse; // CellIdx must match with sdr::ElemSparse, to change, change it there
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: FYI I had to bind connections::CellIdx and sdr::ElemSparse, where ElemSparse is the element type of SDR_sparse_t (currently UInt32); 
Both have to be the same for some methods in SP, TM, so I binded them together. It could be untangeled but then I'd have to do type-conversion of the vectors in said methods.

File: src/nupic/algorithms/Connections.hpp [hunk 17]
```diff
@@ -333,7 +335,7 @@ class Connections : public Serializable
    *
    * @retval A vector length
    */
-  UInt32 segmentFlatListLength() const;
+  size_t segmentFlatListLength() const { return segments_.size(); };
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: implemented `size_t` for Connections, TM, and SP.

File: src/nupic/types/Sdr.hpp [hunk 18]
```diff
@@ -27,17 +27,21 @@
...
 
 namespace nupic {
 namespace sdr {
 
-typedef std::vector<Byte>               SDR_dense_t;
-typedef std::vector<UInt>               SDR_sparse_t;
-typedef std::vector<std::vector<UInt>>  SDR_coordinate_t;
-typedef std::function<void()>           SDR_callback_t;
+using ElemDense        = Byte; //TODO allow changing this
+using ElemSparse       = UInt32; //must match with connections::CellIdx
```
Comment (breznak) [inline src/nupic/types/Sdr.hpp]: ElemSparse & ElemDense represent data type of elements in SDR_sparse/dense_t; 
- ElemSparse is used to match in Connections with CellIdx
- ElemDense is in preparation for changing this (at runtime with teplate?)

**Comment by breznak** [comment]
@dkeeney @ctrl-z-9000-times please have a look when you have time, I'd like to continue tomorrow and it'd help to have this already in.

File: src/nupic/algorithms/Connections.hpp [hunk 19]
```diff
@@ -375,21 +376,12 @@ class Connections : public Serializable
...
-  void
-  computeActivity(std::vector<SynapseIdx> &numActiveConnectedSynapsesForSegment,
-                  const std::vector<CellIdx> &activePresynapticCells) const;
-
-  void
-  computeActivity(std::vector<SynapseIdx> &numActiveConnectedSynapsesForSegment,
-                  std::vector<SynapseIdx> &numActivePotentialSynapsesForSegment,
-                  const CellIdx activePresynapticCell,
-                  const Permanence connectedPermanence) const;
+  void computeActivity(std::vector<SynapseIdx> &numActiveConnectedSynapsesForSegment,  //TODO remove the 2 arg version, use only the 3 arg (with 1 being optional)
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.hpp]: No can do, the two different versions do different things.  One computes less than have as much as the other, making it significantly faster.  Removing it would likely double the spatial poolers runtime.

File: src/nupic/algorithms/Connections.hpp [hunk 20]
```diff
@@ -449,41 +441,46 @@ class Connections : public Serializable
...
+  size_t numCells() const { return cells_.size(); }
 
   /**
    * Gets the number of segments.
    *
    * @retval Number of segments.
    */
-  Segment numSegments() const;
+  size_t numSegments() const { 
+	  NTA_ASSERT(segments_.size() >= destroyedSegments_.size()); //TODO remove destroyedSegments_?
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.hpp]: This assert is not needed.
destroyedSegments_ is very much needed, otherwise segments could never be destroyed.

File: src/nupic/algorithms/Connections.hpp [hunk 21]
```diff
@@ -449,41 +441,46 @@ class Connections : public Serializable
...
+  size_t numSegments(CellIdx cell) const { return cells_[cell].segments.size(); }
 
   /**
    * Gets the number of synapses.
    *
    * @retval Number of synapses.
    */
-  Synapse numSynapses() const;
+  size_t numSynapses() const {
+    NTA_ASSERT(synapses_.size() >= destroyedSynapses_.size());
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.hpp]: nit: assert not needed.

File: src/nupic/algorithms/Connections.hpp [hunk 22]
```diff
@@ -449,41 +441,46 @@ class Connections : public Serializable
...
-  SynapseIdx numSynapses(Segment segment) const;
+  size_t numSynapses(Segment segment) const { return segments_[segment].synapses.size(); }
 
   /**
    * Comparison operator.
    */
-  bool operator==(const Connections &other) const;
-  bool operator!=(const Connections &other) const;
+  virtual bool operator==(const Connections &other) const;
+  inline bool operator!=(const Connections &other) const { return !operator==(other); }
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Connections.hpp]: `virtual` is not needed, since this class should never be subclassed.  It's not supported, this is the only virtual method here. I expect if it were subclassed that a lot of things would break.

My understanding is that all modern compilers ignore `inline` and figuring out what to inline on its own.

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 23]
```diff
@@ -268,35 +270,29 @@ static void growSynapses(Connections &connections, Random &rng, Segment segment,
...
-      nActual, maxSynapsesPerSegment - connections.numSynapses(segment));
+  const size_t nActualWithMax = std::min(nActual, static_cast<size_t>(maxSynapsesPerSegment) - connections.numSynapses(segment));
 
   // Pick nActual cells randomly.
-  for (UInt32 c = 0; c < nActualWithMax; c++) {
-    UInt32 i = rng.getUInt32((UInt32)candidates.size());
-    connections.createSynapse(segment, candidates[i], initialPermanence);
+  for (size_t c = 0; c < nActualWithMax; c++) {
+    const auto i = rng.getUInt32(static_cast<UInt32>(candidates.size()));
+    connections.createSynapse(segment, candidates[i], initialPermanence); //TODO createSynapse create a vector of new synapses at once
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/TemporalMemory.cpp]: I disagree with this TODO note.  I don't see any advantage to making a list of synapses and then making them all at once.  It's more work & complexity to make, populate, and then iterate over an extra list.  I think the performance of this issue is likely insignificant either way.

The next line however is almost certainly a performance issue:
`candidates.erase(candidates.begin() + i);`

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 24]
```diff
@@ -460,7 +456,10 @@ static void punishPredictedColumn(
 }
 
 void TemporalMemory::activateCells(const SDR &activeColumns, bool learn) {
-    NTA_CHECK( activeColumns.dimensions == columnDimensions_ );
+    NTA_CHECK( activeColumns.dimensions.size() == columnDimensions_.size() );
+    for(size_t i=0; i< columnDimensions_.size(); i++) {
+      NTA_CHECK(static_cast<size_t>(activeColumns.dimensions[i]) == static_cast<size_t>(columnDimensions_[i])) << "Dimensions must be the same.";
+    }
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/TemporalMemory.cpp]: Can we keep the shorter version?  Also the comments.
`NTA_CHECK( activeColumns.dimensions == columnDimensions_ ) << "Dimensions must be the same.";`

File: src/nupic/regions/TMRegion.cpp [hunk 25]
```diff
@@ -251,12 +252,13 @@ void TMRegion::compute() {
...
-      active = VectorHelpers::sparse_cellsToColumns<UInt>(active, args_.cellsPerColumn);
-      predictive = VectorHelpers::sparse_cellsToColumns<UInt>(predictive, args_.cellsPerColumn);
+      active = VectorHelpers::sparse_cellsToColumns<CellIdx>(active, args_.cellsPerColumn);
+      predictive = VectorHelpers::sparse_cellsToColumns<CellIdx>(predictive, args_.cellsPerColumn);
     }
     SDR& sdr = out->getData().getSDR();
-    VectorHelpers::unionOfVectors<UInt>(sdr.getSparse(), active, predictive);
-    sdr.setSparse(sdr.getSparse()); // to update the cache in SDR.
+    vector<CellIdx> uni;
+    VectorHelpers::unionOfVectors<CellIdx>(uni, active, predictive);
```
Comment (ctrl-z-9000-times) [inline src/nupic/regions/TMRegion.cpp]: Before this was an in-place update and now it makes a new vector and swaps it in...

File: src/nupic/algorithms/Connections.hpp (hunk 19 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: > One computes less than have as much as the other, making it significantly faster. 

I'm aware, if we use {} the loop will skip, or some empheral value, as in TM(?)

File: src/nupic/algorithms/Connections.hpp (hunk 20 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: - can I keep the assert?, I guess someone can make a working code-change where #destroyed > #segments_ 

> One computes less than have as much as the other, making it significantly faster. 

can we just rm the SegmentIdx from the list of segments?

File: src/nupic/algorithms/Connections.hpp (hunk 22 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: I'll remove the virtual, you're right about the inline, but it does not hurt either.

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 24 above)
Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: > Can we keep the shorter version? 

unfortunately no, typedefs in Connections allow a change (CellIdx, and I'd prefer to keep CellIdx even the type for columnDimensions, as it directly tells the user the range. ) 
- while SDR::dimensions is UInt

- so I was getting unsupported compare `vector<uint32> == vector<uint16>`;
  - if we do element-by-element, NTA_CHECK is able to do the comparison

**Comment by breznak** [comment]
> ake smaller PRs. The size_t changes probably could have had their own PR, and then most of this work would have been merged a day ago. 

- you're right, this PR grew big, justification for adding size_t is that it's needed for the typedef changes to reliably work
   - is it OK to review as is, or do I need to split? 

> Don't leave so many TODO notes in the code. Either do them, or talk about them on this github forum. Every change in the code is fair game for us to argue over, and it can really slow things down. If it's a comment of the github page then we can discuss it without blocking the PR from merging. 

this is true, I've abused the code for discussion here, I'll rm the PRs or make them to issues.

File: src/nupic/algorithms/Connections.hpp (hunk 20 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: > can we just rm the SegmentIdx from the list of segments?

No, because then the other segment indexes would change.  We could go through the work of updating these references since they're mostly contained within the connections class, but I'm not sure it would be any faster.

File: src/nupic/algorithms/Connections.hpp (hunk 22 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: `inline` is a nitpick, you can ignore me here.

**Comment by ctrl-z-9000-times** [comment]
> is it OK to review as is, or do I need to split?

This is fine to review as it is, especially since I've already done the review.  I just meant in general.  Thanks!

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 24 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: Ok, that makes sense.  Maybe in a future PR:  SDR.dimsnsions could use CellIdx too?

File: src/nupic/regions/TMRegion.cpp (hunk 25 above)
Comment (breznak) [reply src/nupic/regions/TMRegion.cpp]: Good spot! 
This was for mismatch in SDR_sparse_t, and vector<CellIdx> //you can change CellIdx; 
but got fixed later by binding connections::CellIdx and sdr::ElemSparse

TL;DR reverted back

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 23 above)
Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: > I don't see any advantage to making a list of synapses and then making them all at once. 

I tend to prefer passing data in vectors as long as possible, it avoids nested loops and is better for spotting hot-spots. Also SIMD (auto)vectorization can do a better job then. 

> The next line however is almost certainly a performance issue:
candidates.erase(candidates.begin() + i);
 
good point! 

Anyway, out of scope of the PR, updated the comments and I'll open a separate issue.

File: src/nupic/algorithms/Connections.hpp (hunk 19 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: #371

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 23 above)
Comment (breznak) [reply src/nupic/algorithms/TemporalMemory.cpp]: #373

**Comment by breznak** [comment]
@ctrl-z-9000-times I hope to have addressed your feedback, one more review, please

**Comment by breznak** [comment]
> This looks great to me! 

thanks for bearning with me on this PR!
